design/open/spark-outputcommitter/committer.md

```diff
@@ -0,0 +1,148 @@
+
+# Why <a href="#user-content-why" id="user-content-why">#</a>
+
+## Issues with existing OutputCommitters
+
+OutputCommitters are charged with writing a directory tree (a multi-part
+Hadoop file, for example a partitioned Parquet file) atomically.
+
+Hadoop comes with several OutputCommitter algorithms.  For S3 these try to
+bridge the gap between the Hadoop assumed semantics of a POSIX-ish
+filesystem (HDFS) and the actual semantics of an object store (S3).
+
+The best correct committer is probably the [magic committer][magic].  It
+works using only S3 (with its current consistency guarantees!), without
+auxiliary systems such as a small HDFS or S3Guard.  The magic committer
+uploads a directory tree by starting a multipart upload to its intended location,
+creating a completion record for that multipart upload on S3, and uploading
+the object without completing the upload.  To complete the upload, all
+completion records are processed and the objects magically appear.
+
+The [staging committer][staging] (aka "Netflix committer") is an older
+committer for S3A.  It requires a "staging" filesystem on HDFS and offers
+manual modes for overwriting outputs ("conflict resolution").
+
+These issues remain with the magic and staging committer:
+
+* Not atomic.  Multipart objects appear sequentially.  Finding a "_success"
+  object indicator is still required before a directory tree can be
+  processed.
+* The magic committer requires "magic committer" support in the FileSystem,
+  in order to write `__magic` objects to a _different_ path.  (Current
+  lakeFSFS does not contain this support.)
+* The magic committer is somewhat new:
+  > It’s also not been field tested to the extent of Netflix’s committer;
+  > consider it the least mature of the committers.
+* Documentation is still somewhat lacking.
+
+Other OutputCommitters are not as good, and _rename by copying_.  This is
+both slow and highly non-atomic, as well as making it difficult to recover
+from failed attempts.  The current recommendations are to use the
+partitioned staging committer for overwriting or updating partitioned data
+trees, and the magic committer is most other cases.
+
+## LakeFSOutputCommitter
+
+We propose to leverage the atomic capabilities of lakeFS to write a specific
+OutputCommitter for lakeFSFS.  In the initial version, it will branch out,
+prepare the desired output, and merge back in to commit.  Aborting will be
+done by dropping the branch (or repurposing it if the same job ID is
+requested again).  Regular lakeFS retention can handle dropping data
+objects; we might want to add file patterns to retention to allow temporary
+objects to be dropped rapidly.
+
+# How
+
+## Sample flow
+
+* User configures lakeFSFS and configures LakeFSOutputCommitter as the
+  OutputCommitter for `lakefs` protocol Paths.  **Possibly** lakeFSFS will
+  set up this OutputCommitter by default.
+* **Setup (job/task TBD)**: Create a new branch for this job/task.  Its name
```

Comment:
Maybe stress that these are operations/functions that are part of the OutputCommitter API.